File: with_setting/167341/10550.py
from collections import Counter

def solution(pouches):

    max_count = 0;
    # todo 1. 낱말을 abcde 문자열로 순회하며 진행
    for flavor in 'abcde':
        # todo 2. contributions 기여도 구하기
        contributions = [];
        for pouch in pouches:
            counter = Counter(pouch)
            this = counter[flavor];
            etc = sum(counter.values()) - this;
            contribution = this - etc;
            contributions.append(contribution);
            contributions.sort(reverse=True);
            
            # todo 3. 기여도의 누적합을 구하기
            length = len(contributions);
            print(f'length: {length}');
            count = 0;
            prefixSum = [0]*length  # prefixSum을 담을 빈 배열 생성
            prefixSum[0] = contributions[0] # 첫번째 값을 넣어줌
            for i in range(1,length): # 두번째 값부터 누적합을 저장
                prefixSum[i] = prefixSum[i-1] + contributions[i]
                if prefixSum[i] <= 0:
                    count += i
                    break;
        
        max_count = max(max_count, count);
            
    return max_count;

pouches = ["cab", "adaaa", "e"] #3 
# pouches = ["aabb", "baba"] #0 
# pouches = ["d", "a", "e", "d", "abdcc"] #3
# pouches = ["a"] #1
ans = solution(pouches)
print(ans)

File: with_setting/167341/10550.java
import java.util.*;
import java.lang.*;
import java.io.*;

// The main method must be in a class named "Main".
class Main {
    public static int solution(String[] pouches) {
        int max_count = 0;
        // todo 1. 낱말을 abcde 문자열로 순회하며 진행
        for (char i = 'a'; i <= 'e'; i++) {
            System.out.println("flavor : "+i);
            // todo 2. contributions 기여도 구하기
            List<Integer> contributions = new ArrayList<>();
            
            for (String pouch : pouches) {
                int thisFlavor = 0;
                for (char c : pouch.toCharArray()) {
                    if (c == i) {
                        thisFlavor += 1;
                    }
                }
                int etcFlavor = pouch.length() - thisFlavor;
                contributions.add(thisFlavor - etcFlavor);
                System.out.println("contributions : "+contributions);

            }
            Collections.sort(contributions, Collections.reverseOrder());

            // todo 3. 기여도 누적합 구하면서 count에 추가하기
            int total = 0;
            int count = 0;
            for (int contribution : contributions) {
                total += contribution;
                if (total > 0) {
                    count++;
                } else {
                    break;
                }
            }
            max_count = Math.max(max_count, count);
        }
        return max_count;
    }
    public static void main(String[] args) {
        // String[] pouches = {"cab", "adaaa", "e"}; // to be 3;
        // String[] pouches = {"aabb", "baba"}; // to be 0;
        String[] pouches = {"d", "a", "e", "d", "abdcc"}; // to be 3;
        // String[] pouches = {"a"}; // to be 1;
        System.out.println(solution(pouches));
    }
}
